File: src/main/java/com/meada/whatsapp/profiles/papelaria/PapelariaFulfillment.java
package com.meada.whatsapp.profiles.papelaria;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

/**
 * Forma de entrega de um pedido papelaria (camada 8.15 / perfil papelaria) — MATERIALIZADO, espelho
 * 1:1 de {@code frontend/profiles/papelaria/papelaria-fulfillment.ts}. Clone de
 * {@link com.meada.whatsapp.profiles.padaria.PadariaFulfillment} (camada 8.8): o cliente escolhe
 * RETIRADA (balcão, sem taxa/endereço) ou ENTREGA (exige endereço + soma a taxa). O
 * {@code PapelariaFulfillmentParityTest} garante a paridade Java↔TS; a CHECK de
 * {@code papelaria_orders.fulfillment} (migration 59) trava os mesmos ids no banco.
 */
public enum PapelariaFulfillment {
    RETIRADA("retirada", "Retirada"),
    ENTREGA("entrega", "Entrega");

    private final String id;
    private final String label;

    PapelariaFulfillment(String id, String label) {
        this.id = id;
        this.label = label;
    }

    public String id() {
        return id;
    }

    public String label() {
        return label;
    }

    /** Resolve uma forma de entrega pelo id estável. Optional vazio se inválido/null. */
    public static Optional<PapelariaFulfillment> fromId(String id) {
        if (id == null) {
            return Optional.empty();
        }
        return Arrays.stream(values()).filter(f -> f.id.equals(id)).findFirst();
    }

    /** Todas as formas de entrega, na ordem de declaração. */
    public static List<PapelariaFulfillment> allActive() {
        return List.of(values());
    }
}

File: frontend/profiles/papelaria/papelaria-fulfillment.ts
/**
 * Forma de entrega de um pedido papelaria (camada 8.15) — espelho 1:1 de
 * src/main/java/com/meada/whatsapp/profiles/papelaria/PapelariaFulfillment.java.
 *
 * O pedido é RETIRADA (o cliente busca na loja) ou ENTREGA (delivery no endereço). O fluxo do
 * Kanban diverge no fim por causa disso: pronto → retirado (retirada) ou pronto → saiu_entrega →
 * entregue (entrega).
 * O PapelariaFulfillmentParityTest (backend) garante que os ids aqui e no enum Java nunca divergem
 * (o teste casa textualmente cada objeto `{ id: '...' }` deste arquivo).
 * A CHECK de papelaria_orders.fulfillment trava os mesmos ids no banco.
 */
export const PAPELARIA_FULFILLMENTS = [
  { id: 'retirada', label: 'Retirada' },
  { id: 'entrega', label: 'Entrega' },
] as const

export type PapelariaFulfillment = (typeof PAPELARIA_FULFILLMENTS)[number]
export type PapelariaFulfillmentId = PapelariaFulfillment['id']

/** Rótulo pt-BR de uma forma de entrega (fallback: o próprio id se desconhecido). */
export function fulfillmentLabel(id: string): string {
  return PAPELARIA_FULFILLMENTS.find((f) => f.id === id)?.label ?? id
}
